pkg/gatewayconfigurationserver/server.go

```diff
@@ -0,0 +1,162 @@
+// Copyright © 2019 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gatewayconfigurationserver
+
+import (
+	"context"
+	"encoding"
+	"encoding/json"
+	"net/http"
+
+	"github.com/gogo/protobuf/types"
+	"github.com/gorilla/mux"
+	"github.com/grpc-ecosystem/grpc-gateway/runtime"
+	"go.thethings.network/lorawan-stack/v3/pkg/component"
+	gcsv2 "go.thethings.network/lorawan-stack/v3/pkg/gatewayconfigurationserver/v2"
+	"go.thethings.network/lorawan-stack/v3/pkg/pfconfig/cpf"
+	"go.thethings.network/lorawan-stack/v3/pkg/pfconfig/semtechudp"
+	"go.thethings.network/lorawan-stack/v3/pkg/ttnpb"
+	"go.thethings.network/lorawan-stack/v3/pkg/web"
+	"go.thethings.network/lorawan-stack/v3/pkg/webhandlers"
+	"go.thethings.network/lorawan-stack/v3/pkg/webmiddleware"
+	"google.golang.org/grpc"
+)
+
+// Server implements the Gateway Configuration Server component.
+type Server struct {
+	*component.Component
+	config *Config
+}
+
+func (s *Server) withGateway(next func(http.ResponseWriter, *http.Request, *ttnpb.Gateway)) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		gtwID := gatewayIDFromContext(ctx)
+		cc, err := s.GetPeerConn(ctx, ttnpb.ClusterRole_ENTITY_REGISTRY, gtwID)
+		if err != nil {
+			webhandlers.Error(w, r, err)
+			return
+		}
+		client := ttnpb.NewGatewayRegistryClient(cc)
+		gtw, err := client.Get(ctx, &ttnpb.GetGatewayRequest{
+			GatewayIdentifiers: gtwID,
+			FieldMask: types.FieldMask{
+				Paths: []string{
+					"antennas",
+					"frequency_plan_id",
+					"gateway_server_address",
+				},
+			},
+		}, s.WithClusterAuth())
+		if err != nil {
+			webhandlers.Error(w, r, err)
+			return
+		}
+		next(w, r, gtw)
+	}
+}
+
+func (s *Server) makeJSONHandler(f func(context.Context, *ttnpb.Gateway) (interface{}, error)) http.HandlerFunc {
+	return s.withGateway(func(w http.ResponseWriter, r *http.Request, gtw *ttnpb.Gateway) {
+		msg, err := f(r.Context(), gtw)
+		if err != nil {
+			webhandlers.Error(w, r, err)
+			return
+		}
+		w.Header().Set("Content-Type", "application/json")
+		w.WriteHeader(http.StatusOK)
+		enc := json.NewEncoder(w)
+		enc.SetIndent("", "\t")
+		enc.Encode(msg)
```

Comment:
Maybe we should add a handler in `webhandlers` that does this?

Also considering the fact that we have stdlib encoding/json, jsonpb and and ttnjsonpb that each have different output.

Reply:
Yeah so precisely for the reason that we have all sorts of output per endpoint, add the indentation factor, and that we step away from a web framework, I decided to just spell this out